### PR TITLE
Write sys params

### PR DESCRIPTION
**Pull request by vtnate**
#### Any background context you want to provide?
First connection between UO SDK & UO DES
#### What does this PR accomplish?
Reads files from UO SDK and writes a system_parameters.json file for use in DES
#### How should this be manually tested?
`py.test tests/geojson/test_csv_to_sys_param.py`
~User/tester needs to run a scenario with the UO SDK first, and edit line 42 of `test_csv_to_sys_param.py` to point to the scenario folder once the run is complete. SDK run needs to include [Amy's measure(s)](https://github.com/urbanopt/DES_HVAC/tree/develop/Measures), which aren't in the SDK yet, which adds another wrinkle to manual testing.~
#### What are the relevant tickets?
#211 
#### Screenshots (if appropriate)

@nllong I've got a couple TODO & FIXME's in here, with further work that needs to be done. This shows my current progress, and shouldn't be too crazy to move into the SystemParameters class if that's where we want this to live.

### CODE REVIEW

**Comment by vtnate** [comment]
Test failures are due to the hardcoded test path that only exists on my local machine.

**Comment by vtnate** [comment]
@nllong tests are ready now.

File: geojson_modelica_translator/geojson/csv_to_sys_param.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    def __init__(self, scenario_dir=None, sys_param_template=None):
+        if Path(scenario_dir).exists():
+            self.scenario_dir = scenario_dir
+            # FIXME: this feature_file can't be hardcoded, name & location need to be user specified
+            self.feature_file = self.scenario_dir.parent.parent / "example_project.json"
+        else:
+            raise Exception(f"Unable to find your scenario. The path you provided was: {scenario_dir}")
+
+        # TODO: sys_param_template can be hardcoded, once we decide where this whole thing lives
+        if Path(sys_param_template).exists():
```
Comment (vtnate) [inline geojson_modelica_translator/geojson/csv_to_sys_param.py:52]: Am I right that we should plan to hard-code the sys_param_template.json file?

File: geojson_modelica_translator/geojson/csv_to_sys_param.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+        # TODO: Thermal zone names - useful for Spawn in the future in separate method
+        # input_loads_columns = list(self.input_loads_file.columns)
+        # thermal_zone_names = []
+        # for column_header in input_loads_columns:
+        #     if 'Zone' in column_header:
+        #         thermal_zone_names.append(column_header.split('_')[-1])
+        # thermal_zone_names = list((set(thermal_zone_names)))
+
+        param_template['buildings']['custom'] = building_list
```
Comment (macintoshpie) [inline geojson_modelica_translator/geojson/csv_to_sys_param.py:129]: should we issue some kind of warning (or error out) if `building_list` is empty after the filtering?

**Review by macintoshpie — COMMENTED** [review]
Everything looks great to me 👍 My review was focused on the python code and running the tests, so hopefully @nllong can provide better feedback regarding GMT requirements.
I had just a few style/general questions and comments.

After running `test_csv_to_sys_param.py` I ended up with this json file, which creates the object under the `custom` key, and updates `nominal_flow_district` with the max value from the `massFlowRateHeating` column in `geojson-modelica-translator/tests/geojson/data/sdk_output_skeleton/run/baseline_15min/7/014_export_time_series_modelica/building_loads.csv`
```json
{
  "buildings": {
    "default": {
      "load_model": "time_series",
      "ets_model": "Indirect Heating and Cooling",
      "ets_model_parameters": {
        "indirect": {
          "q_flow_nominal": 8888,
          "eta_efficiency": 0.9,
          "nominal_flow_district": 2.324,
          "nominal_flow_building": 0.5,
          "pressure_drop_valve": 1000,
          "pressure_drop_hx_secondary": 1500,
          "pressure_drop_hx_primary": 1500,
          "cooling_supply_water_temperature_district": 5,
          "cooling_supply_water_temperature_building": 7,
          "heating_supply_water_temperature_district": 55,
          "heating_supply_water_temperature_building": 50,
          "booster_heater": false,
          "ets_generation": "Fourth Generation"
        }
      }
    },
    "custom": [
      {
        "geojson_id": "7",
        "load_model": "time_series",
        "load_model_parameters": {
          "time_series": {
            "filepath": "<Ted's local path>/geojson-modelica-translator/tests/geojson/data/sdk_output_skeleton/run/baseline_15min/7/015_export_modelica_loads/modelica.mos",
            "delTDisCoo": 10,
            "nominal_flow_building": 2.324
          }
        }
      }
    ]
  }
}
```

**Comment by vtnate** [comment]
> After running `test_csv_to_sys_param.py` I ended up with this json file, which creates the object under the `custom` key, and updates `nominal_flow_district` with the max value from the `massFlowRateHeating` column in `geojson-modelica-translator/tests/geojson/data/sdk_output_skeleton/run/baseline_15min/7/014_export_time_series_modelica/building_loads.csv`
> ```

The `nominal_flow_district` is actually the sum of the max values from `massFlowRateHeating` from each feature, but in this PR there is only one feature. I have added more features to demonstrate this more clearly in #256 .

File: geojson_modelica_translator/geojson/csv_to_sys_param.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+class CSVToSysParam(object):
+    """
+    Parser for URBANopt output to write a system_parameters file
+    """
+
+    def __init__(self, scenario_dir=None, sys_param_template=None):
+        if Path(scenario_dir).exists():
+            self.scenario_dir = scenario_dir
+            # FIXME: this feature_file can't be hardcoded, name & location need to be user specified
+            self.feature_file = self.scenario_dir.parent.parent / "example_project.json"
```
[the remark this answers is not in the document]
Comment (vtnate) [reply geojson_modelica_translator/geojson/csv_to_sys_param.py:47]: No, no reason. Fix coming. Thank you for your review, this is great!

File: geojson_modelica_translator/geojson/csv_to_sys_param.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        # TODO: move all these methods iside the SystemParameters class
+        # sys_param = SystemParameters(sys_param_filename, template)
+        # sys_param.populate_filenames(geojson)
+
+        # Parse the sys_param template
+        with open(self.sys_param_template) as template_file:
+            param_template = json.load(template_file)
+
+        # TODO: get the results in this comprehension instead of using measure_list
+        [self.parse_feature(x) for x in self.scenario_dir.iterdir() if x.is_dir()]
```
[the remark this answers is not in the document]
Comment (vtnate) [reply geojson_modelica_translator/geojson/csv_to_sys_param.py:87]: @macintoshpie Can you help me? I've tried several wrong ways and now I'm stuck in a rut.

**Comment by macintoshpie** [comment]
@vtnate Sorry but I think adding my suggestion for renaming the method breaks your code a bit, you would also need to update the _call_ to the method as well. Feel free to just undo the changes and keep it as it was originally instead, it was intended to be a more general comment about python method naming conventions, I probably should have left out the code suggestion.

**Comment by vtnate** [comment]
> @vtnate Sorry but I think adding my suggestion for renaming the method breaks your code a bit, you would also need to update the _call_ to the method as well. Feel free to just undo the changes and keep it as it was originally instead, it was intended to be a more general comment about python method naming conventions, I probably should have left out the code suggestion.

Ha, yep, I noticed. I'm working on a new commit that handles that.

**Comment by vtnate** [comment]
@macintoshpie The sys_param template is specific to the model type, following the example of [model_connectors](https://github.com/urbanopt/geojson-modelica-translator/tree/develop/tests/model_connectors/data).

File: geojson_modelica_translator/geojson/csv_to_sys_param.py (hunk 1 above)
Comment (vtnate) [reply geojson_modelica_translator/geojson/csv_to_sys_param.py:55]: @macintoshpie The sys_param template is specific to the model type, following the example of [model_connectors](https://github.com/urbanopt/geojson-modelica-translator/tree/develop/tests/model_connectors/data).

**Comment by vtnate** [comment]
@nllong This is ready for your review
@macintoshpie gave an excellent review, resulting in a number of changes and much better code. Questions for Nick:
- Does this create the output file you want?
- Do you want this class/collection-of-methods to live somewhere else?

File: geojson_modelica_translator/geojson/csv_to_sys_param.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+        self.sys_param_template = Path(__file__).parent.parent.parent / 'tests' / 'geojson' / 'data' / 'time_series_template.json'
+
+        self.measure_list = []
+
+    def _parse_items(self, measure_folder: Path) -> None:
+        """
+        Go through each folder (OpenStudio measure) in a feature and pull out the mass-flow-rate data file & modelica loads file
+        """
+        if str(measure_folder).endswith('_export_time_series_modelica'):
```
Comment (nllong) [inline geojson_modelica_translator/geojson/csv_to_sys_param.py:62]: who wrote that OpenStudio workflow gem with this type of structure? 👀 😬 (it was me, sorry.)

File: geojson_modelica_translator/geojson/csv_to_sys_param.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+        """
+        for item in feature.iterdir():
+            if item.is_dir():
+                self._parse_items(item)
+
+    def csv_to_sys_param(self, sys_param_filename, overwrite=True):
+        if Path(sys_param_filename).exists() and not overwrite:
+            raise Exception(f"Output file already exists and overwrite is False: {sys_param_filename}")
+
+        # TODO: move all these methods inside the SystemParameters class
```
Comment (nllong) [inline geojson_modelica_translator/geojson/csv_to_sys_param.py:79]: Yes, this needs to happen. This fork is going to cause a lot of pain. Once we merge this down can you create an issue and start working on it?

File: geojson_modelica_translator/geojson/csv_to_sys_param.py [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+        # sys_param = SystemParameters(sys_param_filename, template)
+        # sys_param.populate_filenames(geojson)
+
+        # Parse the sys_param template
+        with open(self.sys_param_template) as template_file:
+            param_template = json.load(template_file)
+
+        for x in self.scenario_dir.iterdir():
+            if x.is_dir():
+                self._parse_feature(x)
```
Comment (nllong) [inline geojson_modelica_translator/geojson/csv_to_sys_param.py:89]:
```suggestion
        for dir in self.scenario_dir.iterdir():
            if dir.is_dir():
                self._parse_feature(dir)
```

File: geojson_modelica_translator/geojson/csv_to_sys_param.py [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+        with open(self.feature_file) as json_file:
+            data = json.load(json_file)
+            for feature in data['features']:
+                if not feature['properties']['type'] == 'Site Origin':
+                    building_ids.append(feature['properties']['id'])
+
+        # Make sys_param template entries for each feature_id
+        building_list = []
+        for building in building_ids:
+            feature_info = deepcopy(param_template['buildings']['custom'][0])
```
Comment (nllong) [inline geojson_modelica_translator/geojson/csv_to_sys_param.py:102]: this should also end up in the system param class. We access the data in the system parameter file using `jsonpath`.

File: tests/geojson/test_csv_to_sys_param.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+****************************************************************************************************
+"""
+
+import unittest
+from pathlib import Path
+from shutil import rmtree
+
+from geojson_modelica_translator.geojson.csv_to_sys_param import CSVToSysParam
+
```
Comment (nllong) [inline tests/geojson/test_csv_to_sys_param.py:36]: NOICE, thanks for the tests!

File: tests/geojson/test_csv_to_sys_param.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+class CSVToSysParamTest(unittest.TestCase):
+    def setUp(self):
+        self.data_dir = Path(__file__).parent / 'data'
+        self.output_dir = Path(__file__).parent / 'output'
+        self.scenario_dir = self.data_dir / "sdk_output_skeleton" / "run" / "baseline_15min"
```
Comment (nllong) [inline tests/geojson/test_csv_to_sys_param.py:42]:
```suggestion
        self.scenario_dir = self.data_dir / 'sdk_output_skeleton' / 'run' / 'baseline_15min'
```

**Review by nllong — APPROVED** [review]
looks good! I'm going to merge in the CLI first... so hold up.

**Comment by nllong** [comment]
@vtnate -- tests aren't going to pass which we know. Let me know when you finish the review comments and ready to merge! Thanks!

**Pull request merged**